{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "\n",
    "# input ncbi accession number for the genome of interest\n",
    "id = 'NC_005856.1'\n",
    "\n",
    "# input paths to gff and fasta files for the genome of interest\n",
    "path_to_gff = 'gff/'\n",
    "path_to_fasta = 'fasta/'\n",
    "\n",
    "# input PAM sites to be searched for\n",
    "pam_sites = [\"TTTC\", \"TTTG\", \"TTTA\"]\n",
    "ideal_positions_in_gene = [0.2, 0.5] # where 20% downstream of gene start site is 0.2\n",
    "length_to_be_extracted = 28 # length of the sequence to be extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper for get_genes\n",
    "def get_product(str, to_find):\n",
    "    pos1 = str.find(to_find)\n",
    "    pos2 = str.find(';', pos1, len(str))\n",
    "    return str[pos1 + len(to_find):pos2]\n",
    "\n",
    "# opens gff3 file, uses it to access every gene's id, start position, end position, name, and strand (in that order)\n",
    "def get_genes(path_id):\n",
    "    f = open(path_id + '.gff3', 'r')\n",
    "    genes = []\n",
    "    for line in f.readlines():\n",
    "        l = line.split('\\t')\n",
    "        if len(l) > 4 and l[2] == 'gene':\n",
    "            genes.append((get_product(l[8], 'ID=gene-'), int(l[3]), int(l[4]), get_product(l[8], 'Name='), l[6]))\n",
    "        elif len(l) > 4 and id == 'EF056009.1' and l[2] == 'CDS':\n",
    "            genes.append((get_product(l[8], 'product='), int(l[3]), int(l[4]), get_product(l[8], 'Name='), l[6]))\n",
    "    return genes\n",
    "\n",
    "# opens fasta file, accesses seqs of all of the genes + strand directions\n",
    "def get_seqs(s1e2, path_id):\n",
    "    seq = SeqIO.read(path_id + '.fasta', 'fasta')\n",
    "    gene = []\n",
    "    for i in range(len(s1e2)):\n",
    "        if s1e2[i][4] == \"-\":\n",
    "#            print('negative')\n",
    "            gene.append(str((seq.seq[s1e2[i][1] : s1e2[i][2]]).reverse_complement()))\n",
    "        elif s1e2[i][4] == \"+\":\n",
    "            gene.append(str(seq.seq[s1e2[i][1] : s1e2[i][2]]))\n",
    "#            print('positive')\n",
    "        else:\n",
    "            print('strand direction not identifiable:')\n",
    "            print(s1e2[i][4])\n",
    "    return gene\n",
    "\n",
    "# used to determine whether the PAM site is viable\n",
    "# current requirements: \n",
    "def passes_tests(a_str, start):\n",
    "    if start >= len(a_str) - length_to_be_extracted:\n",
    "        return False\n",
    "    t = extract_primer(a_str, start)\n",
    "    if 'GAAGAC' in t or 'GTCTTC' in t:\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "# for a given gene sequence, find as many entries of sites_to_look_for\n",
    "# as possible, and compile them into a list of potential primer sites\n",
    "def find_primer_candidates(gene_seq, sites_to_look_for):\n",
    "    gene_seq = gene_seq.lower()\n",
    "    start = 0\n",
    "    while True:\n",
    "        min_start = 999999999\n",
    "        min_idx = -1\n",
    "        for i in range(len(sites_to_look_for)):\n",
    "            possible_start = gene_seq.find(sites_to_look_for[i].lower(), start)\n",
    "            if possible_start > -1:\n",
    "                min_start = min(possible_start, min_start)\n",
    "                min_idx = i\n",
    "        if min_start == 999999999: return\n",
    "        start = min_start\n",
    "        if passes_tests(gene_seq, start):\n",
    "            yield start\n",
    "        start += (len(sites_to_look_for[min_idx]) + length_to_be_extracted)\n",
    "        \n",
    "# given the gene sequence, extract the primer immediately\n",
    "# downstream of the identified primer location\n",
    "def extract_primer(seq, pam_loc):\n",
    "    return seq[pam_loc+4:pam_loc+4+length_to_be_extracted]\n",
    "\n",
    "# given a set of candidate primer locations (cand), the length of the \n",
    "# corresponding gene (length), and two ideal %downstreams of the gene \n",
    "# in question (loc1 and loc2), find the two primer candidates that \n",
    "# are closest to loc1 and loc2 in terms of how far downstream in the gene\n",
    "# they are\n",
    "def get_indices(cand, loc1, loc2, length):\n",
    "    int1 = int(loc1 * length)\n",
    "    int2 = int(loc2 * length)\n",
    "    \n",
    "    i = 0\n",
    "    not_found = True\n",
    "    while not_found:\n",
    "        for candidate in cand:\n",
    "            if candidate in range(int1 - i, int1 + i):\n",
    "                sol1 = candidate\n",
    "                not_found = False\n",
    "        i += 1\n",
    "        \n",
    "    not_found = True\n",
    "    i = 0\n",
    "    while not_found:\n",
    "        for candidate in cand:\n",
    "            if candidate in range(int2 - i, int2 + i) and candidate is not sol1:\n",
    "                sol2 = candidate\n",
    "                not_found = False\n",
    "        i += 1\n",
    "    \n",
    "    return [sol1, sol2]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name_start_end_locs = get_genes(path_to_gff + id)\n",
    "genes = get_seqs(gene_name_start_end_locs, path_to_fasta + id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the primer candidates for each set, as well as how many there are for each gene in question\n",
    "primer_candidates = []\n",
    "for i in range(len(genes)):\n",
    "    primer_candidates.append(list(find_primer_candidates(genes[i],[\"TTTC\", \"TTTG\", \"TTTA\"])))\n",
    "count = [len(primer_start_pos_list) for primer_start_pos_list in primer_candidates]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene 3, P1_gp005, had only one possible primer candidate sequence.\n",
      "gene 12, P1_gp115, had only one possible primer candidate sequence.\n",
      "gene 45, P1_gr044, had only one possible primer candidate sequence.\n",
      "gene 71, P1_gt070, had no possible primer candidate sequences.\n",
      "gene 72, P1_gt071, had only one possible primer candidate sequence.\n",
      "gene 77, P1_gp076, had only one possible primer candidate sequence.\n",
      "gene 78, P1_gt117, had only one possible primer candidate sequence.\n",
      "gene 89, P1_gp087, had only one possible primer candidate sequence.\n",
      "gene 99, P1_gp097, had only one possible primer candidate sequence.\n",
      "gene 100, P1_gp098, had no possible primer candidate sequences.\n",
      "gene 102, P1_gp100, had only one possible primer candidate sequence.\n",
      "gene 108, P1_gp106, had no possible primer candidate sequences.\n",
      "gene 115, P1_gp113, had only one possible primer candidate sequence.\n",
      "gene 116, P1_gp114, had no possible primer candidate sequences.\n"
     ]
    }
   ],
   "source": [
    "# for each gene, first determine how many primer candidates there are.\n",
    "# if 0: indicate to the user that this gene has no primer candidates. Can't do much else here.\n",
    "# if 1: create a fwd and rev primer out of that primer candidate (called <gene name>_1_FWD and <gene name>_1_REV)\n",
    "#       and add it to the list of primers\n",
    "# if 2: find the best two candidates with get_indices(); create fwd and rev primers out of these two candidates\n",
    "primers = []\n",
    "for i in range(len(genes)):\n",
    "    if count[i] > 1:\n",
    "        indices = get_indices(primer_candidates[i], ideal_positions_in_gene[0], ideal_positions_in_gene[1], len(genes[i]))\n",
    "        prime1 = Seq(extract_primer(genes[i],indices[0]))\n",
    "        primer1_fwd = 'AGAT' + str(prime1).lower() + 'G'\n",
    "        primer1_rev = 'GAAAC' + str(prime1.reverse_complement()).lower()\n",
    "        prime2 = Seq(extract_primer(genes[i],indices[1]))\n",
    "        primer2_fwd = 'AGAT' + str(prime2).lower() + 'G'\n",
    "        primer2_rev = 'GAAAC' + str(prime2.reverse_complement()).lower()\n",
    "        primers.append([gene_name_start_end_locs[i][0] + '_1_FWD', primer1_fwd, gene_name_start_end_locs[i][0] + '_1_REV', primer1_rev, round(indices[0] / len(genes[i]), 2)])\n",
    "        primers.append([gene_name_start_end_locs[i][0] + '_2_FWD', primer2_fwd, gene_name_start_end_locs[i][0] + '_2_REV', primer2_rev, round(indices[1] / len(genes[i]), 2)])\n",
    "    elif count[i] == 1:\n",
    "        prime1 = Seq(extract_primer(genes[i],primer_candidates[i][0]))\n",
    "        primer1_fwd = 'AGAT' + str(prime1).lower() + 'G'\n",
    "        primer1_rev = 'GAAAC' + str(prime1.reverse_complement()).lower()\n",
    "        primers.append([gene_name_start_end_locs[i][0] + '_1_FWD', primer1_fwd,gene_name_start_end_locs[i][0] + '_1_REV', primer1_rev, round(primer_candidates[i][0] / len(genes[i]),2)])\n",
    "        print('gene ' + str(i) + ', ' + gene_name_start_end_locs[i][0] + ', had only one possible primer candidate sequence.')\n",
    "    else:\n",
    "        print('gene ' + str(i) + ', ' + gene_name_start_end_locs[i][0] + ', had no possible primer candidate sequences.')\n",
    "        primers.append([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(id + '_primers' + '.csv','w') as result_file:\n",
    "    wr = csv.writer(result_file, dialect='excel')\n",
    "    wr.writerows(primers)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
